File: src/vault.js
/*
 * Backup framework
 *
 * Copyright (C) 2012, 2013 Jolla Ltd.
 * Contact: Denis Zalevskiy <denis.zalevskiy@jollamobile.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.

 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 * http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
 */

require("qtcore.js");
require("json2.js");

var sys = require("sys.js");
var os = require("os.js");
var subprocess = require("subprocess.js");
var util = require("util.js");
var error = require("error.js");
var debug = require("debug.js");
var git = require("git.js");

Date.method('toGitTag', function() {
    return this.toISOString().replace(/:/g, '-');
});

var mk_snapshots = function(vcs) {
    var id = function(name) { return '>' + name; }
    , is_id = function(id) { return id.length && id[0] === '>'; }
    , name = function(id) {
        return is_id(id) ? id.substr(1) : undefined;
    }
    , that = Object.create({
        list : function() {
            return util.map(util.filter(vcs.tags(), is_id), name);
        },
        note : function(treeish) { return vcs.notes.get(treeish); },
        activate : function(name) { vcs.checkout(id(name)); },
        tag : function(name) { vcs.tag([id(name)]); }
    });
    return that;
};

var mk_vault = function(path) {

    var vcs = git(path);
    var storage = os.path(path, ".git");
    var blob_storage = os.path(storage, 'blobs');
    var message_file = os.path(path, ".message");
    var snapshots = mk_snapshots(vcs);

    var init = function(config) {
        config["status.showUntrackedFiles"] = "all";

        if (!os.mkdir(path))
            error.raise({
                msg : "Can't init vault",
                path : path,
                reason : "directory already exists" });

        try {
            if (vcs.init())
                error.raise({
                    msg : "Can't init git",
                    path : path,
                    stderr : vcs.stderr()});

            if (!os.path.exists(storage))
                error.raise({
                    msg : "Can't find .git",
                    path : path,
                    stderr : vcs.stderr()});

            vcs.set_config(config);
            var anchor_file = os.path(path, '.vault');
            os.write_file(anchor_file, sys.date().toGitTag());
            vcs.add(anchor_file);
            vcs.commit('anchor');
            vcs.tag(['anchor']);
            os.path.isdir(blob_storage) || os.mkdir(blob_storage);
        } catch (err) {
            os.rmtree(path);
            throw err;
        }
    };

    var status_dump = function(status) {
        return util.map(status, function(item) {
            return item.toString();
        }).join(', ');
    };

    var tag_as_latest = function() {
        vcs.tag(['-d', '>latest'], true);
        vcs.tag(['>latest']);
    };

    /**
     * load or initialize vault configuration describing
     * registered backup modules. Configuration is read-only,
     * mutable() method returns object to modify it
     */
    var mk_config = function() {
        var config_fname = ".config";
        var config_path = os.path(path, config_fname);
        var config, res;
        config = (os.path.isfile(config_path)
                  ? JSON.parse(os.read_file(config_path))
                  : {});

        var save = function() {
            os.write_file(config_path, JSON.stringify(config, null, '\t'));
        };

        res = {};

        res.modules = function() { return config; };

        /// create wrapper to change configuration
        res.mutable = function() {
            var that = Object.create(res);
            that.method('add', function(desc) {
                var name = desc.name;
                if (!(name && desc.script))
                    error.raise({
                        msg : "Module description should contain"
                            + " name and script"});
                config[name] = desc;
                save();
                vcs.add(config_fname);
                vcs.commit("+" + name);
            });
            that.method('rm', function(name) {
                if (name && (name in config)) {
                    delete config[name];
                    save();
                    vcs.rm(config_fname);
                    vcs.commit("-" + name);
                } else {
                    error.raise({
                        msg : "Can't delete non-existing module",
                        name : name });
                }
            });
            return that;
        };

        return res;
    };

    /// lazy vault configuration loading/initialization
    var init_config = function() {
        return mk_config();
        // if (modules_config == null)
        //     modules_config = mk_config()
        // return modules_config
    };

    var blob = function(git_path) {
        var sha = vcs.hash_object(git_path)
        , prefix = sha.slice(0, 2)
        , id = sha.slice(2)
        , blob_dir = os.path(blob_storage, prefix)
        , blob_fname = os.path(blob_dir, id)
        , link_fname = os.path(vcs.root(), git_path);

        var add = function() {
            if (os.path.isfile(blob_fname)) {
                os.unlink(link_fname);
            } else {
                os.rename(link_fname, blob_fname);
            }
            os.symlink(os.path.relative
                       (blob_fname, os.path.dirname(link_fname))
                       , link_fname);
        };

        return Object.create({
            add : add
        });
    };

    /// functionality related to specific module
    var mk_module = function(config, home) {
        var name = config.name;
        var root_dir = vcs.path(name);
        var data_dir = root_dir.path("data");
        var blobs_dir = root_dir.path("blobs");
        var mkdir = os.mkdir;

        var reset = function(treeish) {
            vcs.clean(['-fd', '--', name]);
            vcs.reset(['--hard', treeish]);
        };

        /// execute backup script registered for the module
        var exec_script = function(action) {
            debug.debug('script', config.script, 'action', action);
            var args = ['--action', action,
                        '--dir', data_dir,
                        '--bin-dir', blobs_dir,
                        '--home-dir', home ];
            debug.info(subprocess.check_output(config.script, args));
        };

        var restore = function() {
            exec_script('import');
        };

        var backup = function() {
            var status, i;

            var is_tree_dirty = function(status) {
                return util.first(status, function(item) {
                    return !item.isTreeClean();
                }) < status.length;
            };

            var isnt_commited = function(status) {
                return util.first(status, function(item) {
                    return !item.isClean();
                }) < status.length;
            };

            // cleanup directories for data and blobs in
            // the repository
            os.rmtree(data_dir.absolute);
            os.rmtree(blobs_dir.absolute);
            mkdir(root_dir.absolute);
            mkdir(data_dir.absolute);
            mkdir(blobs_dir.absolute);

            exec_script('export');

            // save blobs
            util.foreach(vcs.status(blobs_dir.relative), function(status) {
                var git_path = status.src;
                if (status.index == ' ' && status.tree == 'D')
                    return vcs.rm(git_path);

                return blob(git_path).add();
            });

            // commit data
            status = vcs.status(root_dir);
            if (!status.length) {
                debug.info("Nothing to backup for " + name);
                return;
            }

            vcs.add(root_dir, ['-A']);
            status = vcs.status(root_dir);
            if (is_tree_dirty(status))
                error.raise({msg : "Dirty tree",
                             dir : root_dir,
                             status : status_dump(status) });

            vcs.commit(">" + name);

            status = vcs.status(root_dir);
            if (isnt_commited(status))
                error.raise({msg : "Not fully commited",
                             dir : root_dir,
                             status : status_dump(status)});

        };
        return Object.create
        ({ backup : backup,
           restore : restore,
           reset : reset });
    };

    var backup = function(home, options, on_progress) {
        var res = { succeeded :[], failed : [] };
        var config = init_config();
        var start_time_tag = sys.date().toGitTag();
        var name, message;

        var backup_module = function(name) {
            var head_before = vcs.rev_parse('HEAD');
            //config[name].name = name
            var module = mk_module(config.modules()[name], home);

            try {
                on_progress({ module: name, status: "begin" });
                module.backup();
                on_progress({ module: name, status: "ok" });
                res.succeeded.push(name);
            } catch (err) {
                err.module = name;
                debug.error("Failed to backup " + name + ", reason: "
                            + err.toString());
                on_progress({ module: name, status: "fail" });
                res.failed.push(name);
                module.reset(head_before);
            }
        };

        vcs.checkout('master');

        if (options && options.modules) {
            util.foreach(options.modules, backup_module);
        } else {
            for (name in config.modules()) {
                backup_module(name);
            }
        }

        message = ((options && options.message)
                   ? [start_time_tag, options.message].join('\n')
                   : start_time_tag);
        os.write_file(message_file, message);
        vcs.add(".message");
        vcs.commit([start_time_tag, message].join('\n'));

        vcs.snapshots.tag(start_time_tag);
        tag_as_latest();
        vcs.notes.add(options.message || start_time_tag);
        return res;
    };

    var restore = function(home, options, on_progress) {
        var config = init_config();
        var res = { succeeded :[], failed : [] };
        var name;

        var restore_module = function(name) {
            //config[name].name = name
            var module = mk_module(config.modules()[name], home);
            try {
                on_progress({ module: name, status: "begin" });
                module.restore();
                on_progress({ module: name, status: "ok" });
                res.succeeded.push(name);
            } catch (err) {
                err.module = name;
                debug.error("Failed to restore " + name
                            + ", reason: " + err.toString());
                on_progress({ module: name, status: "fail" });
                res.failed.push(name);
            }
        };

        if (options && options.modules) {
            util.foreach(options.modules, restore_module);
        } else {
            for (name in config.modules()) {
                restore_module(name);
            }
        }
    };

    return Object.create({
        /// init vault git repository
        init : init,
        /// perform backup
        backup : backup,
        restore : restore,
        snapshots : snapshots,
        /// returns repository configuration
        config : mk_config,
        checkout : function(treeish) { vcs.checkout(treeish) }
    });
};

var parse_kv_pairs = function(cfg) {
    var res = {};
    var pairs, i, kv;
    if (cfg) {
        util.foreach(cfg.split(','), function(v) {
            kv = v.split('=');
            if (kv.length == 2 && kv[0].length)
                res[kv[0]] = kv[1];
        });
    }
    return res;
};

var results = (function() {
    var that = function(obj) {
        var dst = (obj.status === 'ok'
                   ? that.succeeded
                   : that.failed);
        dst.push(obj.module);
    };

    that.succeeded = [];
    that.failed = [];
    return that;
}).call();

mk_vault.execute = function(options) {
    var vault = mk_vault(options.vault);
    var action = options.action;
    var res;

    switch (action) {
    case 'init':
        res = vault.init(parse_kv_pairs(options.git_config));
        break;
    case 'backup':
        res = vault.backup(options.home,
                           {modules : [options.module],
                            message : options.message},
                           results);
        break;
    case 'restore':
        if (!options.tag)
            error.raise({msg : "tag should be provided to restore"});
        vault.snapshots.activate(options.tag);
        res = vault.restore(options.home,
                            {modules : [options.module]},
                            results);
        break;
    case 'list-snapshots':
        res = vault.snapshots.list();
        print(res.join('\n'));
        break;
    case 'register':
        vault.checkout('master');
        if (!options.data)
            error.raise({ action : action, msg : "Needs data" });
        res = vault.config().mutable().add(parse_kv_pairs(options.data));
        break;
    case 'unregister':
        vault.checkout('master');
        if (!options.module)
            error.raise({ action : action, msg : "Needs module name" });
        res = vault.config().mutable().rm(options.module);
        break;
    default:
        error.raise({ msg : "Unknown action", action : action});
        break;
    }
    return res;
};

return mk_vault;

File: src/vault.cpp
#include <QCoreApplication>
#include <QString>
#include <QProcess>
#include <QDebug>
#include <QDir>
#include <QScriptEngine>
#include <QSharedPointer>
#include <cor/options.hpp>
#include <cor/error.hpp>
#include <unordered_map>

class Error : public std::runtime_error
{
public:
    Error(QString const &s)
        : std::runtime_error(s.toStdString())
    { }

    virtual void print_trace() const
    {
        for (auto s : trace)
            std::cerr << s << std::endl;
    }

private:
    cor::Backtrace<30> trace;
};

class Process
{
public:
    Process(QString const &cmd, QString const &cwd = "")
        : cmd_(cmd)
    {
        if (cwd != "")
            p_.setWorkingDirectory(cwd);
    }

    int execute(QStringList const &);
    
    QByteArray stdout()
    {
        return p_.readAllStandardOutput();
    }

    QByteArray stderr()
    {
        return p_.readAllStandardError();
    }

    QString cwd() const
    {
        return p_.workingDirectory();
    }
    
private:
    QProcess p_;
    QString cmd_;
};

int Process::execute(QStringList const &args)
{
    p_.start(cmd_, args);
    p_.waitForFinished();
    return p_.exitStatus();
}

template <typename ... Args>
int shell(QString const &cmd, Args... args)
{
    Process p(cmd);
    return p.execute(QStringList({args...}));
}

template <typename T, typename FnT>
void erase_if(QList<T> &src, FnT fn)
{
    for (auto cur = src.begin(); cur != src.end();)
        if (fn(*cur))
            cur = src.erase(cur);
        else
            ++cur;
}

bool mkdir(QString const& path)
{
    bool is_created = false;
    QDir d(path);
    if (!d.exists()) {
        is_created = true;
        auto name = d.dirName();
        if (!d.cdUp())
            throw Error(QString("Parent dir %1 is unaccesible").arg(path));
        if (!d.mkdir(name))
            throw Error(QString("Can't create %1 in %2").arg(name, d.path()));
    }
}

namespace git
{

class Git : public Process
{

public:
    Git(QString const &root)
        : Process("git", root)
    {
    }

    template <typename ... Args>
    int command(QString const &name, Args&... args)
    {
        return execute(QStringList({name, args...}));
    }

    int command(QString const &name, QStringList const &params)
    {
        return execute(QStringList({name}) + params);
    }

    int config(QString const &name, QString const &value)
    {
        return command("config", name, value);
    }

    template <typename ... Args>
    int commit(QString const &msg, Args&... args)
    {
        return command("commit", "-m", msg, args...);
    }

    template <typename ... Args>
    int status(Args&... args)
    {
        return command("status", "-z", args...);
    }

};

class Status
{
public:

    enum Ids {
        Nop = ' ',
        Delete = 'D',
        Modify = 'M',
        Add = 'A',
        Rename = 'R',
        Copy = 'C',
        Unmerge = 'U',
        Unknown = '?',
        Type = 'T'
    };

    Status(Ids idx, Ids tree, QString const &src)
        : idx_(idx), tree_(tree), src_(src)
    { }

    static QList<Status> get(Git &, QString const &path = "");

    bool isBinary() const
    {
        return bin_ops_.count(idx_) || bin_ops_.count(tree_);
    }

    void setDst(QString const &s)
    {
        dst_ = s;
    }

    QString toString() const;
    
    char idx_chr() const { return (char)idx_; }
    char tree_chr() const { return (char)tree_; }

private:

    static std::unordered_map<char, Ids> ids_;
    static std::set<Ids> bin_ops_;

    Ids idx_;
    Ids tree_;
    QString src_;
    QString dst_;
};

std::unordered_map<char, Status::Ids> Status::ids_= {
    {' ', Status::Nop},
    {'D', Status::Delete},
    {'M', Status::Modify},
    {'A', Status::Add},
    {'R', Status::Rename},
    {'C', Status::Copy},
    {'U', Status::Unmerge},
    {'?', Status::Unknown},
    {'T', Status::Type}
};

std::set<Status::Ids> Status::bin_ops_ = {
    Status::Rename,
    Status::Copy
};

QString Status::toString() const
{
    return (dst_.isEmpty()
            ? QString("%1 (%2, %3)").arg
            (src_, (QChar)idx_, (QChar)tree_)
            : QString("%1 -> %2 (%3, %4)").arg
            (src_, dst_, (QChar)idx_, (QChar)tree_));
}

QList<Status> Status::get(Git &ps, QString const &path)
{
    int rc = (path != ""
              ? ps.status("--", path)
              : ps.status());
    if (rc) {
        qDebug() << "Error on status" << rc << "\n" << ps.stderr();
        throw cor::Error("status");
    }
    QList<Status> res;
    QList<QByteArray> items(ps.stdout().split('\0'));
    erase_if(items, [](QByteArray const &v) {return v.length() == 0; });
    for (auto cur = items.begin(); cur != items.end(); ++cur) {
        auto &v = *cur;
        if (v.length() < 4 && v[2] != ' ') {
            qDebug() << "Unexpected status item format:" << v;
            throw cor::Error("Parsing status");
        }
        Status s(ids_[v[0]], ids_[v[1]], v.mid(3));
        if (s.isBinary()) {
            ++cur;
            if (cur == items.end()) {
                qDebug() << "No dst after bin op chunk";
                throw cor::Error("No dst after bin op chunk");
            }
            s.setDst(*cur);
        }
        res.append(s);
    }
    return res;
}

}


class VaultFactory
{
public:
    VaultFactory(QString const &path)
        : initialized_(false),
          git_(new git::Git(path)),
          dir_(git_->cwd())
    {
    }
    ~VaultFactory()
    {
        if (!initialized_ && dir_.exists())
            shell("rm", "-rf", git_->cwd());
    }
    QSharedPointer<git::Git> create()
    {
        mkdir(git_->cwd());
        init();

        initialized_ = true;
        return git_;
    }
private:
    void init()
    {
        static auto err = [this](QString const &msg) {
            throw Error(QString
                        ("Dir %1: %2\nstderr:\n%3")
                        .arg(git_->cwd(), msg, QString(git_->stderr())));
        };
        if (dir_.exists(".git")) {
            if (git_->status())
                err("Invalid repository status");
        } else {
            if (git_->command("init"))
                err("Can't init repository");

            git_->config("user.email", "email");
            git_->config("user.name", "user");
            git_->config("status.showUntrackedFiles", "all");
        }
    }

    bool initialized_;
    QSharedPointer<git::Git> git_;
    QDir dir_;
};

QSharedPointer<git::Git> init_vault(QString const &path)
{
    VaultFactory f(path);
    return f.create();
}

int main(int argc, char *argv[])
{
    typedef cor::OptParse<QString> option_parser_type;
    option_parser_type::map_type opts;
    std::vector<char const*> params;
    option_parser_type options
        ({{'c', "config"}, {'H', "home"}, {'V', "vault"}, {'a', "action"}},
         {{"config", "config"}, {"home", "home"}, {"action", "action"}},
         {"config", "home", "action", "vault"});
    options.parse(argc, argv, opts, params);
    
    std::cout << "Vault " << opts["vault"] << std::endl;
    QCoreApplication app(argc, argv);
    QScriptEngine engine;
    qDebug << engine.evaluate("2 +4").toNumber();
    auto g = init_vault(opts["vault"]);
    auto s = git::Status::get(*g, ".");
    for (auto &v : s)
        qDebug() << v.toString();
    
    return 0;//app.exec();
}
